### ISSUE TITLE
proposal: vector abstraction

### ISSUE THREAD

**Issue by mmcloughlin** [mmcloughlin/avo#71]
Posting proposal from @egonelbre @stuartcarnie in #performance room on Gopher slack.

> I was imagining the ability to define a loop that should be vectorized and avo would generate the necessary SIMD instructions. I think there would be a few different scenarios though, depending one whether we are reducing the input data, such as a sum vs performing some read / write using the multiple lanes of the SIMD register.

> Maybe it’s about defining SIMD ops behind an abstraction that then can be specialized for different SIMD widths.

Egon put together a "mockup" of what this might look like: https://play.golang.org/p/J5Yohon9UYn

```go
package main

func main(){
    n := Load(Param("n"), GP64())
    src := Load(Param("src"), GP64())
    dst := Load(Param("dst"), GP64())

    a := WideRegister()
    b := WideRegister()

    Label("loop")
    {
        LoadWide(src)
        LoadWide(dst)
        AddWide(a, b)
        StoreWide(a, dst)

        CMPQ(n, wreg.Width)
        JGT(LabelRef("loop"))
    }

    ...
}
```

Perhaps we could gain inspiration from C++ SIMD abstraction layers such as https://github.com/p12tic/libsimdpp.

**Comment by stuartcarnie** [comment]
This seems like the right approach 👍  Presumably these abstractions would exist in their own package?

**Comment by mmcloughlin** [comment]
> This seems like the right approach 👍 Presumably these abstractions would exist in their own package?

Yes I'd agree this belongs as another package that uses avo under the hood.
